src/Streamly/Prelude.hs

```diff
@@ -750,6 +770,23 @@ mapM f m = go (toStream m)
             yield a r = S.runStream (toStream (f a |: (go r))) svr stp sng yld
          in (S.runStream m1) Nothing stp single yield
 
+-- | Monadic version mapMaybe.
+--
+-- @since 0.2.1
+{-# INLINE mapMaybeM #-}
+mapMaybeM :: (IsStream t, MonadAsync m) => (a -> m (Maybe b)) -> t m a -> t m b
+mapMaybeM f m = go (toStream m)
+  where
+    go m1 = fromStream $ Stream $ \svr stp sng yld ->
+        let single a = f a >>= \mbb -> case mbb of
+                Just b  -> sng b
+                Nothing -> stp
+            reconstructStream x = S.runStream (toStream x) svr stp sng yld
+            yield a r = f a >>= \mbb -> case mbb of
+                Just b  -> reconstructStream $ return b |: go r
```

Comment:
There is a little problem with this, it has limited parallelism. The problem is that we have to first run the action i.e. `f a >>= \mbb -> case mbb of` before we can compose its result in parallel with `go r`. This means we can never run more than one action in parallel.

The root of the problem is because we are doing map and filter in one step. It will have full parallelism if we implement it as `map fromJust . filter isJust . mapM f`. In that case we can run as many actions in parallel as we want (see `mapM`, it runs the actions in parallel) and then filter in one go.

Reply:
1) When you use `f a` to make a decision it means you need to evaluate the monadic action`f a` to get its result, so the action has already been executed there.  Later, in the `Just` case, we just add the pure value `b`, which is the result of `f a`, to the stream. The `runStream` function now passes this stream constructed with the result of `f a` and the remaining stream `go r` to the continuations.

2) `fmap` is deliberately not concurrent that's why it discards `svr`. fmap is a pure function and therefore not relatively expensive. Only monadic functions are concurrent because monadic actions may be expensive. However, if we want to use a concurrent map for even a pure function we can always use `mapM (return . f)` instead of `fmap f`. In mapMaybeM, the expensive monadic operation is done via concurrent `mapM` but the cheaper `fromJust` is done non-concurrently using `fmap`.

Reply:
Thanks for your patience! Now I understand why my previous versions do not support concurrent execution. I pushed the latest change below using `mapM (return . fromJust)` as you suggested. Now the chain only have `S.mapM` and `S.filter`, which should support concurrent execution.